src/base/kaldi-error.h

```diff
@@ -63,60 +66,50 @@ inline int32 GetVerboseLevel() { return g_kaldi_verbose_level; }
 /// automatically from ParseOptions.
 inline void SetVerboseLevel(int32 i) { g_kaldi_verbose_level = i; }
 
-// Class KaldiLogMessage is invoked from the  KALDI_WARN, KALDI_VLOG and
-// KALDI_LOG macros. It prints the message to stderr.  Note: we avoid
-// using cerr, due to problems with thread safety.  fprintf is guaranteed
-// thread-safe.
-
-// class KaldiWarnMessage is invoked from the KALDI_WARN macro.
-class KaldiWarnMessage {
- public:
-  inline std::ostream &stream() { return ss_; }
-  KaldiWarnMessage(const char *func, const char *file, int32 line);
-  ~KaldiWarnMessage();
- private:
-  std::ostringstream ss_;
+/// Log message severity and source location info.
+struct LogMessageEnvelope {
+  // Positive number cast to LogLevel is for verbose logging.
+  enum Severity {
+    Error = -2,
+    Warning = -1,
+    Info = 0,
+  };
+  // A Severity, or a positive number indicating verbosity level.
+  int severity;
+  const char *func;
+  const char *file;
+  int32 line;
 };
 
-// class KaldiLogMessage is invoked from the KALDI_LOG macro.
-class KaldiLogMessage {
- public:
-  inline std::ostream &stream() { return ss_; }
-  KaldiLogMessage(const char *func, const char *file, int32 line);
-  ~KaldiLogMessage();
- private:
-  std::ostringstream ss_;
-};
-
-// Class KaldiVlogMessage is invoked from the KALDI_VLOG macro.
-class KaldiVlogMessage {
- public:
-  KaldiVlogMessage(const char *func, const char *file, int32 line,
-                   int32 verbose_level);
-  inline std::ostream &stream() { return ss_; }
-  ~KaldiVlogMessage() { fprintf(stderr, "%s\n", ss_.str().c_str()); }
- private:
-  std::ostringstream ss_;
-};
+/// Type of user-provided logging function.
+typedef void (*LogHandler)(const LogMessageEnvelope& envelope,
+                           const char* message);
 
+/// Set logging handler. If called with a non-NULL function pointer, the
+/// function pointed by it is called to send messages to a caller-provided
+/// log. If called with NULL pointer, restores default Kaldi error logging to
+/// stderr.  SetLogHandler is obviously not thread safe.
+LogHandler SetLogHandler(LogHandler);
 
-// class KaldiErrorMessage is invoked from the KALDI_ERROR macro.
-// The destructor throws an exception.
-class KaldiErrorMessage {
- public:
-  KaldiErrorMessage(const char *func, const char *file, int32 line);
+// Class MessageLogger is invoked from the KALDI_ERR, KALDI_WARN, KALDI_LOG and
+// KALDI_LOG macros. It formats the message, then either prints it to stderr or
+// passes to the log custom handler if provided, then, in case of the error,
+// throws an std::runtime_exception.
+//
+// TODO(kkm): Is this still true? C++ standard is very explicit on cerr safety.
+// Note: we avoid using cerr, due to problems with thread safety. fprintf is
```

Comment:
Ah, right, if I'm to believe [this answer](http://stackoverflow.com/a/14637716/1149924) which specifically says "unlike in C++03." I'll update this comment.